### PR TITLE
Image package enhancements

### PR DESCRIPTION
**Pull request by cclerget**
**Description of the Pull Request (PR):**

This PR adds partitions support for image package and remove SIF code redundancies in code base. This is the first step for image package enhancements

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](../CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](../CHANGELOG.md) if necessary according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](../CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Review by mem — COMMENTED** [review]
LGTM

**Review by mem — APPROVED** [review]

File: internal/pkg/image/reader.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	"io"
+	"os"
+)
+
+type readerError string
+
+func (e readerError) Error() string { return string(e) }
+
+const (
+	// ErrNoSection corresponds to a missing CNI configuration path
```
Comment (bauerm97) [inline internal/pkg/image/reader.go]: I think maybe there is a bad copy paste here ;)

`missing CNI configuration path`

File: internal/pkg/image/reader.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	ErrNoSection = readerError("no section found")
+	// ErrNoPartition corresponds to a missing CNI plugin path
```
Comment (bauerm97) [inline internal/pkg/image/reader.go]: Same here

`missing CNI plugin path`

File: internal/pkg/image/image.go [hunk 3]
```diff
@@ -37,18 +37,25 @@ type format interface {
...
-	Writable bool     `json:"writable"`
-	RootFS   bool     `json:"rootFS"`
+	Path       string    `json:"path"`
+	Name       string    `json:"name"`
+	Type       int       `json:"type"`
+	File       *os.File  `json:"-"`
+	Fd         uintptr   `json:"fd"`
+	Source     string    `json:"source"`
+	Writable   bool      `json:"writable"`
+	Partitions []Section `json:"partitions"`
```
Comment (bauerm97) [inline internal/pkg/image/image.go:73]: Might be nice to add clarification of the difference between `Image.Partitions` and `Image.Sections`. Adding a variable comment would probably do the trick:

```
Partitions []Section `json:"partitions"` // Partitions are [...]
Sections []Section `json:"sections"` // Sections are [...]
```

**Review by bauerm97 — CHANGES_REQUESTED** [review]
Right now have just looked at `internal/pkg/image` code and given thoughts there so far. Also, as I mentioned in slack about this package I would like to see this change:

>Right now in each file `sif.go`, `squashfs.go`, etc... at the top there is `const SQUASHFS = 1` for example
>We should rather have this at the top of `internal/pkg/image/image.go` instead of having those spread out in each file

```
const (
	SQUASHFS = iota
	EXT3
	SANDBOX
	SIF
)
```

File: internal/pkg/runtime/engines/singularity/container.go [hunk 4]
```diff
@@ -322,27 +281,32 @@ func (c *container) setupSessionLayout(system *mount.System) error {
...
 	}
 
 	if overlayEnabled {
 		sylog.Debugf("Attempting to use overlayfs (enable overlay = %v)\n", c.engine.EngineConfig.File.EnableOverlay)
+		if imgObject.Type == image.SIF {
+			err = c.setupSIFOverlay(imgObject, c.engine.EngineConfig.GetWritableImage())
+			if err == nil {
+				return c.setupOverlayLayout(system, sessionPath)
+			}
+			sylog.Warningf("%s", err)
```
Comment (bauerm97) [inline internal/pkg/runtime/engines/singularity/container.go]: We need a more descriptive warning message here:

`sylog.Warningf("While attempting to set up SIFOverlay: %s", err)`

File: internal/pkg/runtime/engines/singularity/container.go [hunk 5]
```diff
@@ -322,27 +281,32 @@ func (c *container) setupSessionLayout(system *mount.System) error {
...
 	}
 
-	if c.engine.EngineConfig.GetWritableImage() && !writableTmpfs {
-		imgObject, err := c.loadImage(c.engine.EngineConfig.GetImage(), true)
-		if err != nil {
-			return err
-		}
+	imgObject, err := c.loadImage(c.engine.EngineConfig.GetImage(), true)
+	if err != nil {
+		return err
```
Comment (bauerm97) [inline internal/pkg/runtime/engines/singularity/container.go]: `return fmt.Errorf("while loading image object: %s", err)`

File: internal/pkg/runtime/engines/singularity/container.go [hunk 6]
```diff
@@ -322,27 +281,32 @@ func (c *container) setupSessionLayout(system *mount.System) error {
...
 
+	if c.engine.EngineConfig.GetWritableImage() && !writableTmpfs {
+		sylog.Debugf("Image is writable, not attempting to use overlay or underlay\n")
 		if imgObject.Type == image.SIF {
-			err = c.setupWritableSIFImage(imgObject, overlayEnabled)
+			err = c.setupSIFOverlay(imgObject, c.engine.EngineConfig.GetWritableImage())
 			if err == nil {
 				return c.setupOverlayLayout(system, sessionPath)
 			}
 			sylog.Warningf("%s", err)
```
Comment (bauerm97) [inline internal/pkg/runtime/engines/singularity/container.go]: I think we should make this more descriptive, in the same way as the other comment

File: internal/pkg/runtime/engines/singularity/container.go [hunk 7]
```diff
@@ -818,14 +762,14 @@ func (c *container) addOverlayMount(system *mount.System) error {
 				ov.AddLowerDir(filepath.Join(dst, "upper"))
 			}
 
-			err = system.Points.AddImage(mount.PreLayerTag, src, dst, "ext3", flags, imageObject.Offset, imageObject.Size)
+			err = system.Points.AddImage(mount.PreLayerTag, src, dst, "ext3", flags, offset, size)
```
Comment (bauerm97) [inline internal/pkg/runtime/engines/singularity/container.go:765]:
```
if err := [...]; err != nil {
    return fmt.Errorf("while [...]: %s", err)
}
```

Also, while we're making changes here we should just apply this to the other clauses in this switch block

File: internal/pkg/image/sif.go [hunk 8]
```diff
@@ -31,7 +34,81 @@ func (f *sifFormat) initializer(img *Image, fileinfo os.FileInfo) error {
...
+				Type:   uint32(desc.Datatype),
+				Name:   desc.GetName(),
+			}
+			img.Sections = append(img.Sections, data)
+		}
+	}
+
 	img.Type = SIF
+
+	// UnloadContainer close image, just want to unmap image
```
Comment (bauerm97) [inline internal/pkg/image/sif.go:104]: Is this comment referencing why we're not using a `sif` package method to close the container? If so, no additional feedback here

**Review by bauerm97 — CHANGES_REQUESTED** [review]

File: internal/pkg/image/sif.go (hunk 8 above)
Comment (cclerget) [reply internal/pkg/image/sif.go:104]: Must keep image file descriptor open, so just release mapped memory

**Review by bauerm97 — APPROVED** [review]

**Pull request merged**